pubtools/pulplib/_impl/model/validate.py

```diff
@@ -26,5 +26,82 @@ def __call__(self, inst, attr, value):
             self.member_validator(inst, attr, elem)
 
 
+@s(frozen=True, slots=True, hash=True)
+class NamedMappingValidator:
+    mapping = ib(
+        validator=validators.deep_mapping(
+            value_validator=validators.is_callable(),
+            key_validator=validators.instance_of(str),
+        ),
+        factory=dict,
+    )
+
+    def __call__(self, inst, attr, value):
+        validators.instance_of(dict)(inst, attr, value)
+        extra_values = set(value.keys() - self.mapping.keys())  # pylint: disable=E1101
+        missing_values = set(
+            self.mapping.keys() - value.keys()  # pylint: disable=E1101
+        )  # pylint: disable=E1101
+        if extra_values:
+            raise ValueError(
+                "Mapping {name} contains extra values {extra}".format(
+                    name=attr.name, extra=extra_values
+                )
+            )
+        if missing_values:
+            raise ValueError(
+                "Mapping {name} contains extra values {missing}".format(
+                    name=attr.name, missing=missing_values
+                )
+            )
+
+        for key, val in value.items():
+            self.mapping[key](inst, attr, val)  # pylint: disable=E1136
+
+    def __repr__(self):
+        return "<named_mapping_validator {mapping!r}>".format(mapping=self.mapping)
+
+
+def named_mapping_validator(mapping):
+    return NamedMappingValidator(mapping)
+
+
+@s(kw_only=True, frozen=True, slots=True)
+class ContainerListValidator(object):
```

Comment:
As the validator is fairly complicated, could we please get at least one test which tries to construct an object with *invalid* input and verifies that this fails?